csharp/src/Google.Protobuf/Reflection/FileDescriptor.cs

```diff
@@ -555,7 +555,13 @@ public override string ToString()
         /// </summary>
         public T GetOption<T>(Extension<FileOptions, T> extension)
         {
-            var value = Proto.Options.GetExtension(extension);
+            var options = Proto.Options;
+            if (options == null)
```

Comment:
actually, looking at the whole options API in protobuf C# from a more high level perspective, we only need to check whether `options == null` because e.g. the FileDescriptor doesn't provide access to the file options (the property FileDescriptor.Proto is internal, so one cannot access to FileDescriptor.Proto.Options).  

This actually has several downsides:
- we need extra logic such as a GetOption methods, instead of just exposing e.g. the FileOptions msg (or other options) on a descriptor and letting the user use GetExtension methods the same way they would do it for any other message with extensions (which is the way reading custom options is done for e.g. C++, Java and python, as documented here: https://developers.google.com/protocol-buffers/docs/proto#customoptions)
- because the "options" messages are not exposed, one cannot access the standard (=the non-extension ones) options at all.

This makes me think that the whole GetOptions() and TryGetOption() API is a bad idea and we should rather focus on how to expose the descriptor.Options property safely, because then we wouldn't need to worry about the GetOptions/TryGetOptions etc. API's at all (they could be removed and everyone could just rely on standard extension APIs).

Reply:
The only way we could expose descriptor.Options safely is by deep cloning it (since all the protos are supposed to be immutable). With these APIs the only thing we need to clone is the value, if we need to clone it at all.

Reply:
Maybe the deep clone isn't really all that problematic. Most projects would only have a few options in place and it seems fair to require users to cache the results they got from e.g. messageDescriptor.Options if they wanted to process a large number of options.
It seems to me that just exposing description.Options (as a defensive copy) would lead to a more logical and straightforward API (with smaller footprint and fewer duplications, and also more similar to what other languages do) - WDYT?

Reply:
Could we make a internal wrapper class that exposes the IExtendableMessage<T> API but only implements the pure functions? Then GetOptions would just return an IExtendableMessage implementation.

Reply:
But you actually want to be able to access e.g. the FileOptions' properties (the "standard" options, such as "java_package" etc.) 

see https://github.com/protocolbuffers/protobuf/blob/4059c61f27eb1b06c4ee979546a238be792df0a4/csharp/src/Google.Protobuf/Reflection/Descriptor.cs#L3955